sdks/java/core/src/main/java/org/apache/beam/sdk/testing/PAssert.java

```diff
@@ -121,6 +124,32 @@ private static String nextAssertionName() {
   private PAssert() {}
 
   /**
+   * A {@link DoFn} that counts the number of successful {@link SuccessOrFailure} in the
+   * input {@link PCollection} and counts them. If a failed {@link SuccessOrFailure} is
+   * encountered, it is counted and immediately raised.
+   */
+  private static final class DefaultConcludeFn extends DoFn<SuccessOrFailure, Void> {
+
+    @ProcessElement
+    public void processElement(ProcessContext c) {
+      SuccessOrFailure e = c.element();
+      if (e.isSuccess()) {
+        PAssert.successCounter.inc();
+      } else {
+        PAssert.failureCounter.inc();
+        throw e.assertionError();
```

Comment:
Yes. We'd talked about how we want pipelines with failed assertions to fail fast rather than continue to run. I can also change this, though. This also means that even if the runner does not support the Metrics API, their pipeline would fail with failed assertions.